core/Pattern.carp

```diff
@@ -1,6 +1,15 @@
 (system-include "carp_pattern.h")
 
 (defmodule Pattern
+  (register-type MatchResult "PatternMatchResult" [start Int, end Int])
+  (defn       match2str [match-res] 
+	(fmt "#<MatchResult start=%d end=%d>" 
+	  (MatchResult.start match-res) 
+	  (MatchResult.end   match-res) ))
+  (implements str Pattern.match2str)
```

Comment:
The convention here would be to do

```clojure
(defmodule MatchResult
  (defn str [match-res]
    ; your code
  )
  (implements str PatternMatchResult.str)
)
```

in a nested module, unless `str` clashes with anything.

Reply:
Thank you for the pointer, updated the sources.

Printing a MatchResult in the REPL works, but printing (Array.MatchResult) does not.
What am I missing?
```
> (info Pattern.MatchResult.refstr)
Pattern.MatchResult.refstr : (Fn [(Ref Pattern.MatchResult a)] String)
  Defined at line 8, column 6 in '/home/ptokremin/contrib_ghub/Carp/core/Pattern.carp'
  Implements: (prn str)
> (info Pattern.MatchResult.str)
Pattern.MatchResult.str : (Fn [Pattern.MatchResult] String)
  Defined at line 16, column 6 in '/home/ptokremin/contrib_ghub/Carp/core/Pattern.carp'
  Documentation: converts a `Pattern.MatchResult` to a string.
  Implements: (prn str)
> (Pattern.match    #"<(.-)>{(.-)}" "Whatever we <do>{cmd1} is going to cause some <trouble>{msg2} no matter...")
(MatchResult start=12 end=22)
> (info Pattern.find-all-matches )
Pattern.find-all-matches : (Fn [(Ref Pattern a), (Ref String b)] (Array Pattern.MatchResult))
  Defined at line 47, column 4 in '/home/ptokremin/contrib_ghub/Carp/core/Pattern.carp'
> (Pattern.find-all-matches     #"<(.-)>{(.-)}" "Whatever we <do>{cmd1} is going to cause some <trouble>{msg2} no matter...")
[Warning] No 'prn' function found with type (Fn [Pattern.MatchResult] String).
carp: Can't find any 'prn' function for member 'Inside array.' of type (Fn [Pattern.MatchResult] String)
CallStack (from HasCallStack):
  error, called at src/ArrayTemplates.hs:634:33 in CarpHask-0.5.1.0-EudeCbDnX8j3pnSP31uN8I:ArrayTemplates
```

Reply:
No, the `prn` function required here is of type `(Fn [Pattern.MatchResult] String)`. You defined one of type `(Fn [&Pattern.MatchResult] String)`.

Reply:
Sorry, I omited that bit for brevity sake. Here's the complete module:
```
(defmodule Pattern

  (register-type MatchResult "PatternMatchResult" [start Int, end Int])
  (defmodule MatchResult
    (defn refmatch2str [ref-matchres] 
      (fmt "(MatchResult start=%d end=%d)" 
        (MatchResult.start ref-matchres) 
        (MatchResult.end   ref-matchres) ))
    (implements str Pattern.MatchResult.refmatch2str)
    (implements prn Pattern.MatchResult.refmatch2str)
    ;
    (defn match2str [matchres] 
      (Pattern.MatchResult.refmatch2str &matchres) )
    (implements str Pattern.MatchResult.match2str)
    (implements prn Pattern.MatchResult.match2str)
  )
  ; ... code
)
```

Functions look ok from the transcript:
```
> (info Pattern.MatchResult.match2str)
Pattern.MatchResult.match2str : (Fn [Pattern.MatchResult] String)
  Defined at line 14, column 6 in '/home/ptokremin/contrib_ghub/Carp/core/Pattern.carp'
  Implements: (prn str)
> (info Pattern.MatchResult.refmatch2str)
Pattern.MatchResult.refmatch2str : (Fn [(Ref Pattern.MatchResult a)] String)
  Defined at line 7, column 6 in '/home/ptokremin/contrib_ghub/Carp/core/Pattern.carp'
  Implements: (prn str)
> (Pattern.match    #"<(.-)>{(.-)}" "Whatever we <do>{cmd1} is going to cause some <trouble>{msg2} no matter...")
(MatchResult start=12 end=22)
> (Pattern.find-all-matches     #"<(.-)>{(.-)}" "Whatever we <do>{cmd1} is going to cause some <trouble>{msg2} no matter...")
[Warning] No 'prn' function found with type (Fn [Pattern.MatchResult] String).
carp: Can't find any 'prn' function for member 'Inside array.' of type (Fn [Pattern.MatchResult] String)
CallStack (from HasCallStack):
  error, called at src/ArrayTemplates.hs:634:33 in CarpHask-0.5.1.0-EudeCbDnX8j3pnSP31uN8I:ArrayTemplates
```

Could this be related to nested modules?

Reply:
No, this seems fine now. This might possibly be due to the old array code still looking for a function named `prn` rather than a function marked as implementing `prn`. Maybe @scolsen (who built the `implements` mechanism) has some insight on this?

Reply:
Ok, the problem is elsewhere. I'll open a separate issue.